docs/best-practices.rst

```diff
@@ -0,0 +1,271 @@
+
+.. code::
+
+    {{ var.json.<variable_name> }}
+
+
+.. note::
+
+    In general, you should not write any complex code outside the tasks. The code outside the tasks runs every time Airflow parses the DAG, which happens every second by default.
+
+
+Testing a DAG
+^^^^^^^^^^^^^
+
+Airflow users should treat DAGs as production level code. The DAGs should have various tests to ensure that it produces expected results.
+You can write a wide variety of tests for a DAG. Let's take a look at some of them.
+
+DAG Loader Test
+---------------
+
+This test should ensure that your DAG does not contain a piece of code that raises error while loading.
+No additional code needs to be written by the user to run this test.
+
+.. code::
+
+ python your-dag-file.py
+
+Running the above command without any error ensures your DAG does not contain any uninstalled dependency, syntax errors, etc. 
+
+You can look into :ref:`Testing a DAG <testing>` for details on how to test individual operators.
+
+Unit tests
+-----------
+
+Unit tests ensure that there is no incorrect code in your DAG. You can write a unit test for your tasks as well as your DAG.
+
+**Unit test for loading a DAG:**
+
+.. code::
+
+ from airflow.models import DagBag
+ import unittest
+
+ class TestHelloWorldDAG(unittest.TestCase):
+    def setUp(self):
+        self.dagbag = DagBag()
+
+    def test_dag_loaded(self):
+        dag = self.dagbag.get_dag(dag_id='hello_world')
+        self.assertDictEqual(self.dagbag.import_errors, {})
+        self.assertIsNotNone(dag)
+        self.assertEqual(len(dag.tasks), 1)
+
+**Unit test for custom operator:**
+
+.. code::
+
+ import unittest
+ from airflow.utils.state import State
+
+ class MyCustomOperatorTest(unittest.TestCase):
+    def setUp(self):
+        self.dag = DAG(TEST_DAG_ID, schedule_interval='@daily', default_args={'start_date' : DEFAULT_DATE})
+        self.op = MyCustomOperator(
+            dag = self.dag,
+            task_id='test',
+            prefix='s3://bucket/some/prefix',
+        )
+        self.ti = TaskInstance(task=self.op, execution_date=DEFAULT_DATE)
+
+    def test_execute_no_trigger(self):
+        self.ti.run(ignore_ti_state=True)
+        self.assertEqual(self.ti.state, State.SUCCESS)
+        #Assert something related to tasks results
+
+Self-Checks
+------------
+
+You can also implement checks in a DAG to make sure the tasks are producing the results as expected.
+As an example, if you have a task that pushed data to S3, you can implement a check in the next task. The check should 
+make sure that the partition is created in S3 and check if the data is correct or not.
+
+Similarly, if you have a task that starts a microservice in Kubernetes or Mesos, you should check if the service has started or not using :class:`airflow.sensors.http_sensor.HttpSensor`.
+
+.. code::
+
+ task = PushToS3(...)
+ check = S3KeySensor(
+    bucket_key="s3://bucket/key/foo.parquet"
+ )
+ task.set_downstream(check)
```

Comment:
Please use the `>>` style, not the function style.

And can you move this up one paragraph.